templates/templates.go

```diff
@@ -0,0 +1,418 @@
+package templates
+
+import (
+	"embed"
+	"fmt"
+	permission1 "github.com/ystv/web-auth/infrastructure/permission"
+	"github.com/ystv/web-auth/permission"
+	"github.com/ystv/web-auth/permission/permissions"
+	"github.com/ystv/web-auth/role"
+	"github.com/ystv/web-auth/user"
+	"html/template"
+	"io"
+	"log"
+	"strings"
+	"time"
+)
+
+//go:embed *.tmpl
+var tmpls embed.FS
+
+type (
+	Templater struct {
+		Permission *permission.Store
+		Role       *role.Store
+		User       *user.Store
+	}
+
+	Template     string
+	TemplateType int
+)
+
+const (
+	ForgotTemplate       Template = "forgot.tmpl"
+	NotFound404Template  Template = "404NotFound.tmpl"
+	ForgotEmailTemplate  Template = "forgotEmail.tmpl" // generated by go generate
+	InternalTemplate     Template = "internal.tmpl"
+	LoginTemplate        Template = "login.tmpl"
+	NotificationTemplate Template = "notification.tmpl"
+	ResetTemplate        Template = "reset.tmpl"
+	ErrorTemplate        Template = "error.tmpl"
+	SettingsTemplate     Template = "settings.tmpl"
+	SignupTemplate       Template = "signup.tmpl"
+	UserTemplate         Template = "user.tmpl"
+	UsersTemplate        Template = "users.tmpl"
+	RolesTemplate        Template = "roles.tmpl"
+	RoleTemplate         Template = "role.tmpl"
+	ResetEmailTemplate   Template = "resetEmail.tmpl" // generated by go generate
+	PermissionsTemplate  Template = "permissions.tmpl"
+	PermissionTemplate   Template = "permission.tmpl"
+
+	NoNavType      TemplateType = 0
+	PaginationType TemplateType = 1
+	RegularType    TemplateType = 2
+)
+
+func NewTemplate(p *permission.Store, r *role.Store, u *user.Store) *Templater {
+	return &Templater{
+		Permission: p,
+		Role:       r,
+		User:       u,
+	}
+}
+
+func (t Template) String() string {
+	return string(t)
+}
+
+func (t *Templater) RenderTemplate(w io.Writer, data interface{}, mainTmpl Template, templateType TemplateType) error {
+	var err error
+
+	t1 := template.New("_base.tmpl")
+
+	t1.Funcs(t.getFuncMaps())
+
+	switch templateType {
+	case NoNavType:
+		t1 = template.New("_baseNoNavs.tmpl")
+		t1.Funcs(t.getFuncMaps())
+		t1, err = t1.ParseFS(tmpls, "_baseNoNavs.tmpl", "_head.tmpl", "_footer.tmpl", string(mainTmpl))
+	case PaginationType:
+		t1, err = t1.ParseFS(tmpls, "_base.tmpl", "_head.tmpl", "_footer.tmpl", "_navbar.tmpl", "_sidebar.tmpl", "_pagination.tmpl", mainTmpl.String())
+	case RegularType:
+		t1, err = t1.ParseFS(tmpls, "_base.tmpl", "_head.tmpl", "_footer.tmpl", "_navbar.tmpl", "_sidebar.tmpl", mainTmpl.String())
+	default:
+		return fmt.Errorf("unable to parse template, invalid type: %d", templateType)
+	}
+
+	t1.Funcs(t.getFuncMaps())
+
+	if err != nil {
+		log.Printf("failed to get templates for template(RenderTemplate): %+v", err)
+		return err
+	}
+
+	return t1.Execute(w, data)
+}
+
+func (t *Templater) RenderEmail(emailTemplate Template) *template.Template {
+	return template.Must(template.New(emailTemplate.String()).ParseFS(tmpls, emailTemplate.String()))
```

Comment:
Does this need to panic, or is there something sensible we can do if this errors?

Reply:
No it doesn't need to panic